search/src/search.rs

```diff
@@ -65,63 +90,77 @@ impl Search {
             let page: u32 = cur_page.try_into().unwrap_or_else(|_| {
                 panic!("Fatal: Field \"page\" ({}) is bigger than u32!", cur_page)
             });
+            let search_result = SearchResult::new(page, segment_ord, doc_id)
+                .map_err(|_e| SearchError(String::from("Failed creating search result!")))?;
             results
                 .entry(cur_title.to_string())
-                .and_modify(|pages| pages.push_back(page))
-                .or_insert_with(|| LinkedList::from([page]));
+                .and_modify(|pages| pages.push_back(search_result))
+                .or_insert_with(|| LinkedList::from([search_result]));
         }
         Ok(results)
     }
 
-    pub fn get_preview(&self, text: String, phrase: &String) -> Result<String> {
-        let p_index = text
-            .to_lowercase()
-            .find(&(phrase.to_lowercase()))
-            .expect("Searched word not found on page!");
-        let text = format!(
-            "{}**{}**{}",
-            &text[0..p_index],
-            &text[p_index..p_index + phrase.len()],
-            &text[p_index + phrase.len()..]
+    pub fn get_preview(&self, search_result: &SearchResult, input: &str) -> Result<String> {
+        println!(
+            "Calling `get_preview` for: query: {} and page: {}",
+            input, search_result.page
         );
-        let start = if p_index > 50 { p_index - 50 } else { 0 };
-        let end = if (p_index + phrase.len() + 50) < text.len() {
-            p_index + phrase.len() + 50
-        } else {
-            text.len()
-        };
-        let preview = &text[start..end];
-        Ok(format!("...{}...", preview))
-    }
+        // Prepare creating snippet.
+        let searcher = self.reader.searcher();
+        let query_parser = QueryParser::for_index(self.index.index(), self.schema.default_fields());
+        let query = query_parser
+            .parse_query(input)
+            .map_err(|e| SearchError(e.to_string()))?;
+        let snippet_generator = SnippetGenerator::create(&searcher, &*query, self.schema.body)
+            .map_err(|e| SearchError(e.to_string()))?;
+        let retrieved_doc = searcher
+            .doc(DocAddress {
+                segment_ord: (search_result.segment_ord),
+                doc_id: (search_result.doc_id),
+            })
+            .map_err(|e| SearchError(e.to_string()))?;
+
+        // Get text on given page
+        let path = retrieved_doc
+            .get_first(self.schema.path)
+            .expect("Fatal: Field \"path\" not found!")
```

Comment:
We should return an error instead and leave the handling to the caller's context

Reply:
Agreed! What is the correct way to do this when `map_err` is not available?

Reply:
I think that's because it is an option, not a result.

https://stackoverflow.com/a/61299534